### PR TITLE
Add DataImportCron status conditions

### PR DESCRIPTION
**Pull request by arnongilboa**
Signed-off-by: Arnon Gilboa <agilboa@redhat.com>

**What this PR does / why we need it**:

The `DataImportCron` controller updates the status conditions in a controlled `DataImportCron` and its managed `DataSource`.

DataImportCron:
- UpToDate - indicates if the the most recent import is successful and `DataSource` is up-to-date. Updated to False whenever the source digest (latest sha256) is updated.
- Progressing - indicates whether the cron is currently in the process of importing. Updated to True if there is a current import and its `DataVolume` is `ImportInProgress`, otherwise False.

DataSource:
- Ready - indicates that the corresponding pvc exists and is populated. Update according to `DataImportCron.Status.LastImportedPVC` `DataVolume`'s `DataVolumeReady` condition, if the `DataVolume` exists. Otherwise False. Unlike `DataImportCron` `UpToDate` condition, this one does not care about newer source digest.

**Release note**:
```release-note
NONE
```

### CODE REVIEW

File: pkg/controller/dataimportcron-controller.go [hunk 1]
```diff
@@ -218,20 +219,27 @@ func (r *DataImportCronReconciler) update(ctx context.Context, dataImportCron *c
 		now := metav1.Now()
 		dataImportCron.Status.LastExecutionTimestamp = &now
 
-		if dataVolume.Status.Phase == cdiv1.Succeeded {
+		switch dataVolume.Status.Phase {
```
Comment (mhenriks) [inline pkg/controller/dataimportcron-controller.go:222]: There is an ImportFailed phase.  Not sure if/how it is currently used, may want to check to see if it should be handled here

File: pkg/controller/dataimportcron-controller.go [hunk 2]
```diff
@@ -218,20 +219,27 @@ func (r *DataImportCronReconciler) update(ctx context.Context, dataImportCron *c
...
 				return res, err
 			}
+		case cdiv1.ImportScheduled:
+			updateDataImportCronCondition(dataImportCron, cdiv1.DataImportCronProgressing, corev1.ConditionFalse, "Import is scheduled", scheduled)
+		case cdiv1.ImportInProgress:
+			updateDataImportCronCondition(dataImportCron, cdiv1.DataImportCronProgressing, corev1.ConditionTrue, "Import is progressing", inProgress)
 		}
 	}
 
+	if err := r.updateDataSource(ctx, dataImportCron); err != nil {
```
Comment (mhenriks) [inline pkg/controller/dataimportcron-controller.go:239]: Looks like some new datasource management was added to this PR in addition to the conditions?

File: pkg/controller/dataimportcron-controller.go [hunk 3]
```diff
@@ -460,6 +484,16 @@ func addDataImportCronControllerWatches(mgr manager.Manager, c controller.Contro
 	}); err != nil {
 		return err
 	}
+	// Watch only for DataSource deletion
+	if err := c.Watch(&source.Kind{Type: &cdiv1.DataSource{}},
```
Comment (mhenriks) [inline pkg/controller/dataimportcron-controller.go:488]: It is possible for delete events to get lost.  If it is critical to do something when a resource is deleted, use a finalizer

File: pkg/controller/dataimportcron-controller.go [hunk 4]
```diff
@@ -195,11 +198,9 @@ func getCronRegistrySource(cron *cdiv1.DataImportCron) (*cdiv1.DataVolumeSourceR
...
 }
 
-// FIXME: update all conditions
 func (r *DataImportCronReconciler) update(ctx context.Context, dataImportCron *cdiv1.DataImportCron) (reconcile.Result, error) {
 	log := r.log.WithName("update")
 	res := reconcile.Result{}
-	//FIXME: add ImportStatus DataVolumePhase so we'll handle only on transition (to Succeeded etc.)
 	importSucceeded := false
 	imports := dataImportCron.Status.CurrentImports
 	if imports != nil {
```
Comment (mhenriks) [inline pkg/controller/dataimportcron-controller.go:206]: I think conditions should be added/updated all the time, not just when `imports != nil`

**Review by mhenriks — COMMENTED** [review]
Would be helpful if pr description is updated to include conditions/values/transitions.

File: pkg/controller/dataimportcron-controller.go (hunk 4 above)
Comment (arnongilboa) [reply pkg/controller/dataimportcron-controller.go:206]: The `DataImportCronUpToDate` condition is already updated all the time. When `imports == nil` the missing condition update is `DataImportCronProgressing` = false. Adding it. Thanks.

File: pkg/controller/dataimportcron-controller.go (hunk 2 above)
Comment (arnongilboa) [reply pkg/controller/dataimportcron-controller.go:239]: Not much, just re-creation of DataSource if somehow deleted, and refactoring of updateDataSource so DataSource will be k8s-updated when its conditions are updated.

File: pkg/controller/dataimportcron-controller.go (hunk 3 above)
Comment (arnongilboa) [reply pkg/controller/dataimportcron-controller.go:488]: As discussed, we won't use a finalizer for DataSource, but (re-)create it in whenever we see it's missing.

File: pkg/controller/dataimportcron-controller.go (hunk 1 above)
Comment (arnongilboa) [reply pkg/controller/dataimportcron-controller.go:222]: Can be added, but I have no idea how to cause the importer pod to fail. When pod is CrashLoopBackOff the DV phase will still be ImportInProgress.

File: tests/dataimportcron_test.go [hunk 5]
```diff
@@ -133,14 +133,20 @@ var _ = Describe("DataImportCron", func() {
...
 				Expect(err).ToNot(HaveOccurred())
-				return cron.Status.LastImportedPVC != nil && cron.Status.LastImportedPVC.Name == currentImportDv
+				progressCond := controller.FindDataImportCronConditionByType(cron, cdiv1.DataImportCronProgressing)
+				upToDateCond := controller.FindDataImportCronConditionByType(cron, cdiv1.DataImportCronUpToDate)
+				return progressCond != nil && progressCond.Status == corev1.ConditionFalse &&
+					upToDateCond != nil && upToDateCond.Status == corev1.ConditionTrue &&
+					cron.Status.LastImportedPVC != nil && cron.Status.LastImportedPVC.Name == currentImportDv
 			}, dataImportCronTimeout, pollingInterval).Should(BeTrue())
 		}
 		if checkGarbageCollection {
```
Comment (mhenriks) [inline tests/dataimportcron_test.go:152]: looking light on tests?  should test at minimum that retention policies work as expected, datasource/datavolume recreated if deleted

File: pkg/controller/dataimportcron-controller.go (hunk 3 above)
Comment (awels) [reply pkg/controller/dataimportcron-controller.go:520]: So we are fine if the delete event gets lost, and it gets recreated on the next reconcile loop?

File: pkg/controller/dataimportcron-controller.go [hunk 6]
```diff
@@ -242,10 +255,20 @@ func (r *DataImportCronReconciler) update(ctx context.Context, dataImportCron *c
 	}
 
 	desiredDigest := dataImportCron.Annotations[AnnSourceDesiredDigest]
-	if desiredDigest != "" && (imports == nil || (importSucceeded && desiredDigest != imports[0].Digest)) {
-		if err := r.createImportDataVolume(ctx, dataImportCron); err != nil {
-			return res, err
+	digestUpdated := desiredDigest != "" && (len(imports) == 0 || desiredDigest != imports[0].Digest)
+	if digestUpdated {
```
Comment (awels) [inline pkg/controller/dataimportcron-controller.go:259]: There are 4 permutations here, in the unit tests below I see only 2 tests. Can we add some unit tests that ensure that the corner cases are doing what they are supposed to do?

File: pkg/controller/dataimportcron-controller.go [hunk 7]
```diff
@@ -101,12 +103,14 @@ func (r *DataImportCronReconciler) initCron(ctx context.Context, dataImportCron
 	if dataImportCron.Annotations == nil {
 		dataImportCron.Annotations = make(map[string]string)
 	}
+	if !HasFinalizer(dataImportCron, dataImportCronFinalizer) {
+		AddFinalizer(dataImportCron, dataImportCronFinalizer)
```
[the remark this answers is not in the document]
Comment (arnongilboa) [reply pkg/controller/dataimportcron-controller.go:107]: Sure, removed the `HasFinalizer()` check, however we update the CR anyway with `LastExecutionTimestamp` in each reconcile, so I see no reason to check if the finalizes changed.

File: pkg/controller/dataimportcron-controller.go (hunk 1 above)
Comment (arnongilboa) [reply pkg/controller/dataimportcron-controller.go:223]: Removing the ImportFailed case, as it's a dead code.

File: pkg/controller/dataimportcron-controller.go (hunk 3 above)
Comment (arnongilboa) [reply pkg/controller/dataimportcron-controller.go:520]: Yes, we found that to be the reasonable behavior.

File: pkg/controller/dataimportcron-controller.go [hunk 8]
```diff
@@ -302,22 +322,32 @@ func (r *DataImportCronReconciler) updateDataSourceOnSuccess(ctx context.Context
...
-			Name:       dataImportCron.Name,
-			UID:        dataImportCron.UID,
-		})
-	}
-	sourcePVC := &cdiv1.DataVolumeSourcePVC{
-		Namespace: dataImportCron.Namespace,
-		Name:      dataImportCron.Status.CurrentImports[0].DataVolumeName,
+	dataSourceCopy := dataSource.DeepCopy()
+
+	if dataSource.Labels == nil {
```
[the remark this answers is not in the document]
Comment (arnongilboa) [reply pkg/controller/dataimportcron-controller.go:322]: @awels For all the CRs created here (CronJob, Job, DataSouce, DVs...)? Also add to new DataImportCron? I don't see you add it to created DVs in its controller.

File: tests/dataimportcron_test.go (hunk 5 above)
Comment (arnongilboa) [reply tests/dataimportcron_test.go:162]: Tests added. Note that currently `RetentionPolicy` specifies whether the `DataVolumes` and `DataSources` are retained when their `DataImportCron` is deleted. However, when `DataVolume` or `DataSource` are deleted during their `DataImportCron` lifetime, they are recreated regardless of the `RetentionPolicy`.

File: pkg/controller/dataimportcron-controller.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (akalenyu) [reply pkg/controller/dataimportcron-controller.go:322]: Yes, we have https://github.com/kubevirt/containerized-data-importer/blob/main/pkg/controller/config-controller.go#L118
It would make sense to label DataSources, DataImportCrons, CronJob/Jobs, DVs (that are created for corn purpose)

[the remark this answers is not in the document]
Comment (arnongilboa) [reply pkg/controller/dataimportcron-controller.go]: Added

**Review by awels — COMMENTED** [review]
/lgtm

File: pkg/controller/dataimportcron-controller.go (hunk 3 above)
Comment (mhenriks) [reply pkg/controller/dataimportcron-controller.go:512]: The question is when will that next reconcile happen?
